File: src/leetcode75/data_structures/binary_tree.ts
import TreeNode from './tree_node';

// Simple Binary Tree data structure based on TreeNode
class BinaryTree {
  root: TreeNode | null;

  constructor(node_values: number[]) {
    this.root = this.build(node_values);
  }

  private build(values: (number | null)[]): TreeNode | null {
    if (!values || values.length === 0) return null;

    const rootNode = new TreeNode({ val: values[0]! });

    const queue: (TreeNode | null)[] = [rootNode];

    let index = 1;

    while (index < values.length) {
      const currentNode = queue.shift();

      if (currentNode) {
        if (values[index] !== null) {
          currentNode.left = new TreeNode({ val: values[index]! });

          queue.push(currentNode.left);
        }

        index++;

        if (index >= values.length) break;

        if (values[index] !== null) {
          currentNode.right = new TreeNode({ val: values[index]! });

          queue.push(currentNode.right);
        }
        index++;
      }
    }

    return rootNode;
  }

  toArray(): (number | null)[] {
    if (!this.root) return [];

    const arr: (number | null)[] = [];
    const queue: (TreeNode | null)[] = [this.root];

    while (queue.length > 0) {
      const currentNode = queue.shift();

      if (currentNode) {
        arr.push(currentNode.val);
        queue.push(currentNode.left);
        queue.push(currentNode.right);
      } else {
        if (queue.some((node) => node !== null)) {
          arr.push(null);
        }
      }
    }

    return arr;
  }
}

export default BinaryTree;

File: src/data_structures/binary_tree.rb
# frozen_string_literal: true

require_relative 'tree_node'

# Simple Binary Tree data structure based on TreeNode
class BinaryTree

  attr_reader :root

  def initialize(node_values)
    @root = build_tree(node_values)
  end

  def build_tree(values)
    return nil if values.nil? || values.empty?

    root_node = TreeNode.new(values[0]) # first element is root

    queue = [root_node]

    index = 1

    while index < values.length
      current_node = queue.shift

      unless values[index].nil?
        current_node.left = TreeNode.new(values[index]) # second element is left node

        queue.push(current_node.left)
      end

      index += 1

      break if index >= values.length

      unless values[index].nil?
        current_node.right = TreeNode.new(values[index]) # second element is right node

        queue.push(current_node.right)
      end

      index += 1
    end

    root_node
  end

  def to_a
    return [] if root.nil?

    arr = []
    queue = [root]

    until queue.empty?
      current_node = queue.shift

      if current_node
        arr << current_node.val

        queue.push(current_node.left)
        queue.push(current_node.right)
      elsif queue.any? { |node| !node.nil? }
        arr << nil
      end
    end

    arr
  end

end
